src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,121 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final int BIT_SHIFT_VALUE = 1;
+    private static final int ONE_ELEMENT_SIZE = 1;
+    private static final int FIRST_ELEMENT_INDEX = 0;
+    private static final int NO_SUCH_ELEMENT_INDEX = -1;
+    private static final String NO_SUCH_ELEMENT_MESSAGE = "There is no element"
+            + " with index %d in list.";
+    private static final String INDEX_OUT_OF_BOUNDS_MESSAGE = "Index %d does not exist in list.";
+
+    private Object[] data;
+    private int size;
+
+    public ArrayList() {
+        data = new Object[DEFAULT_CAPACITY];
+    }
+
+    private Object[] grow() {
+        data = Arrays.copyOf(data, data.length + (data.length >> BIT_SHIFT_VALUE));
+        return data;
+    }
+
+    private int getIndexByElement(T element) {
+        for (int i = 0; i < size; i++) {
+            if (data[i] == element || (data[i] != null && data[i].equals(element))) {
```

Comment:
don't forget to extract expression into a separate variable when you use it multiple times